datamol/align.py

```diff
@@ -0,0 +1,252 @@
+        _template = dm.to_mol(template)
+    elif copy:
+        _template = dm.copy_mol(template)
+    else:
+        _template = template
+
+    if _template is None:
+        return _mol
+
+    if remove_confs:
+        _mol.RemoveAllConformers()
+
+    # EN: to make this more general and robust, we need to first check whether the template
+    # has 2D coordinates or not. If it does not, we need to compute them.
+    # This is a very rare edge case if requests are comming from the UI, but better to check than not xD
+    if _template.GetNumConformers() == 0:
+        _template = compute_2d_coords(_template)
+
+    # EN: now we can align the molecule to the template
+    # but first, we should avoid MCS as much as possible, because it's expensive
+    # so if the template is a subgraph of the molecule, no need to perform any MCS
+    # Another reason for this, is to avoid inconsistency in alignment between molecules that are
+    # supergraph of the template vs molecules that are subgraph of the template.
+    pattern = _template
+    if not _mol.HasSubstructMatch(_template):
+
+        pattern = None
+        mcs_smarts = dm.find_mcs([_mol, _template])
+
+        if mcs_smarts is not None:
+            pattern = dm.from_smarts(mcs_smarts)
+
+    if pattern is not None:
+
+        if auto_select_coord_gen:
+            rdDepictor.SetPreferCoordGen(use_depiction)
+
+        # we would need to compute 2d coordinates for the molecules if it doesn't have any
+        if _mol.GetNumConformers() == 0:
+            _mol = compute_2d_coords(_mol)
+
+        if use_depiction:
+            rdDepictor.GenerateDepictionMatching2DStructure(
+                _mol,
+                reference=_template,
+                refPatt=pattern,
+                acceptFailure=True,
+                allowRGroups=True,
+            )
+        else:
+            query_match = _mol.GetSubstructMatch(pattern)
+            template_match = _template.GetSubstructMatch(pattern)
+            rdMolAlign.AlignMol(_mol, _template, atomMap=list(zip(query_match, template_match)))
+
+    return _mol
+
+
+def auto_align_many(
+    mols: Union[Sequence[Mol], pd.Series],
+    partition_method: str = "anon-scaffold",
+    copy: bool = True,
+    cluster_cutoff: float = 0.7,
+    **kwargs,
+):
+    """Partition a list of molecules into clusters sharing common scaffold of common core,
+    then align the molecules to that common core. This function will compute the list of
+    smiles/smarts representative of each cluster first.
+
+    The returned molecules will have two properties associated to them:
+
+    - `dm.auto_align_many.cluster_id`: the cluster id of the molecule.
+    - `dm.auto_align_many.core`: the smiles/smarts of the core of the cluster.
+
+    Args:
+        mols: A list of molecules to auto align.
+        partition_method: Partition method to use.
+            One of ['scaffold', 'strip-scaffold', 'anongraph-scaffold', 'cluster' and 'anon-scaffold'].
```

Comment:
we should explain what each of this option are doing.

Reply:
will do